cassovary-core/src/main/scala/com/twitter/cassovary/graph/MemoryMappedDirectedGraph.scala

```diff
@@ -122,190 +159,164 @@ object MemoryMappedDirectedGraph {
-    * The input files must have an identical set of edges, and the first file must be sorted by id1, and the second by id2.  The
-    * graph in binary format is written to the given file.  */
-  def sortedEdgeFilesToGraph(edgeListFileSortedById1: File,
-                             edgeListFileSortedById2: File,
-                             graphFile: File): Unit = {
-    val (outDegrees, inDegrees) = accumulateOutAndInDegrees(edgeListFileSortedById1)
-    System.err.println("finished reading degrees at " + new Date())
-    val nodeCount = outDegrees.size()
-    val out = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(graphFile)))
-    def outDegree(id: Int): Int = outDegrees.get(id)
-    def inDegree(id: Int): Int = inDegrees.get(id)
-    writeHeaderAndDegrees(nodeCount, outDegree, inDegree, out)
-
-    var edgeCount = 0L
-    // Write out-neighbors.  Note that they are already sorted by id1, so we just need to write them directly to the
-    // edge data in the output
-    forEachEdge(edgeListFileSortedById1) { (u, v) =>
-      out.writeInt(v)
-      if (edgeCount % (100*1000*1000) == 0)
-        System.err.println(s"wrote $edgeCount half-edges")
-      edgeCount += 1
-    }
-    // Write in-neighbors.
-    forEachEdge(edgeListFileSortedById2) { (u, v) =>
-      out.writeInt(u)
-      if (edgeCount % (100*1000*1000) == 0)
-        System.err.println(s"wrote $edgeCount half-edges")
-      edgeCount += 1
-    }
-    System.err.println(s"wrote $edgeCount half-edges.  Finished at " + new Date())
-    out.close()
-  }
-}
+    // cumulativeNeighborOffset is the byte offset where neighbor data should be written next.
+    // The out-neighbor data starts after the initial 16 bytes, n+1 Longs for out-neighbors and n+1
+    // Longs for in-neighbors.
+    var cumulativeNeighborOffset = 16L + 8L * (nodeCount + 1) * 2
 
+    var halfEdgeCount = 0L // Each edge (id1, id2) has a half-edge for id1 and a half-edge for id2
 
-// TODO: This is temporary; remove it and possibly replace it with a more robust command line
-// tool for converting graphs to binary format.
-object MemoryMappedDirectedGraphBenchmark {
-  def readAdjacencyListToGraph(graphPath: String): DirectedGraph[Node] = {
-    val filenameStart = graphPath.lastIndexOf('/') + 1
-    val graphDirectory = graphPath.take(filenameStart)
-    val graphFilename = graphPath.drop(filenameStart)
-    println("loading graph:" + graphFilename)
-
-    val reader = new AdjacencyListGraphReader(
-      graphDirectory,
-      graphFilename,
-      new NodeNumberer.IntIdentity(),
-      _.toInt) {
-      override def storedGraphDir: StoredGraphDir = StoredGraphDir.BothInOut
+    // To prevent duplicated code between out-neighbor and in-neighbor writing, iterate over the
+    // neighbor types we write (first Out, then In).
+    for (neighborType <- List(GraphDir.OutDir, GraphDir.InDir)) {
```

Comment:
This code is too long. Is there some way to factor it into smaller functions? Eg. one for reading a chunk into an array, one for writing it to the bufferedOutput, one for writing offsets to fileChannel, and one for the outer loop?